### PR TITLE
Workaround intel/19 OMP internal compiler errors in Trilinos

### PR DESCRIPTION
**Pull request by ndellingwood**
by making m_space_instance a OpenMPInternal* pointer type for intel
compilers of major version <= 19

### CODE REVIEW

**Comment by ndellingwood** [comment]
Workaround to avoid ICE with Trilinos builds showing up in intrepid2 with #5290 , kokkos-kernels perf tests https://github.com/kokkos/kokkos-kernels/issues/1445

**Comment by ndellingwood** [comment]
Will need cherry-pick to develop after approval

**Review by PhilMiller — APPROVED** [review]

**Comment by PhilMiller** [comment]
Yuck. I assume this occurs even with 19.0.5?

**Review by dalg24 — COMMENTED** [review]
Fine by me but this is blocking the support for multiple instance

**Comment by ndellingwood** [comment]
> Yuck. I assume this occurs even with 19.0.5?

@PhilMiller yes and yes (it is yucky, and the ICE occurs with intel/19.0.5)

File: core/src/Kokkos_OpenMP.hpp
Comment (Rombur) [inline core/src/Kokkos_OpenMP.hpp:197]: Could we use a regular shared_ptr instead? Then we would not need to change the rest of the code.

Comment (ndellingwood) [reply core/src/Kokkos_OpenMP.hpp:197]: I tried testing diff below with intel/19.0.5 using release-candidate-3.7.00 (instead of changes in this PR), but ICE still occurs:

```diff
diff --git a/core/src/Kokkos_OpenMP.hpp b/core/src/Kokkos_OpenMP.hpp
index 38a537b..74ca3d5 100644
--- a/core/src/Kokkos_OpenMP.hpp
+++ b/core/src/Kokkos_OpenMP.hpp
@@ -189,7 +189,7 @@ class OpenMP {
   uint32_t impl_instance_id() const noexcept { return 1; }
 
  private:
-  Kokkos::Impl::HostSharedPtr<Impl::OpenMPInternal> m_space_instance;
+  std::shared_ptr<Impl::OpenMPInternal> m_space_instance;
 };
```

**Comment by ndellingwood** [comment]
CI perf test failure with CUDA-11.0-NVCC-C++17 build is spurious, unrelated to changes of this PR
```
24: [  PASSED  ] 35 tests.

24: [  FAILED  ] 1 test, listed below:

24: [  FAILED  ] default_exec.overlap_team_policy

24: 

24:  1 FAILED TEST
```

**Comment by crtrott** [comment]
FYI: I tried all kinds of other things, including a simpler HostSharedPtr version without deleter - that failed.  Even if I just add another pointer as member to OpenMP, and then add a copy constructor it ICEs.

**Review by crtrott — APPROVED** [review]
Lets merge this, but do NOT merge it into develop, instead try and find another way in develop to deal with it - potentially due to use of C++17 we anyway gonna have different code generation. The reproducer in the issue is good, with one more addition to the config file it reproduces the error from source (after config) in <1min on my machine.

**Pull request merged**
